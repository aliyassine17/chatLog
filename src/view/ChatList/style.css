.filter {
    text-align: left;
    padding: 10px 0;
    line-height: 40px;
}
.filter input {
        width: 100px;
        height: 20px;
        padding: 5px;
        margin: 5px;
        font-size: 12px;
        border: solid 0.5px #000;
        z-index: 1;
      
}
.messages {
    font-family: Arial, Helvetica, sans-serif;
}