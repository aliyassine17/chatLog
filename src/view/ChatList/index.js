import React, { Component, Fragment } from  'react';
import { PropTypes } from 'prop-types';
import './style.css';
import ChatMessage from './../ChatMessage';


class ChatList extends Component{
  constructor(props){
    super(props);
    this.state = {
      showRecords: 10
    };
    this.handleChange =  this.handleChange.bind(this);
  }

  handleChange(e){
    const { name,value } = e.target;
    this.setState({
      [name]:value
    });
  }
  render(){

    const { showRecords } = this.state;
    const { chatLog } = this.props;

    if (!chatLog){
      return null;
    }

    let displayPerPage = 10;
    if (!isNaN(showRecords) && showRecords <= chatLog.length ){
      displayPerPage = showRecords;
    }

    const displayList = chatLog.slice(0,displayPerPage);

    const list = displayList.map( chat => (<ChatMessage key={chat.messageId} {...chat} ></ChatMessage>));
    return (
      <Fragment>
        <div className="filter">
          <label htmlFor="showRecords">Records per page </label>
          <input name="showRecords" id="showRecords"
            type="number" value={showRecords}
            onChange={this.handleChange}
            min="1" max={chatLog.length}
          ></input>
        </div>
        <div className="messages">
          {list}
        </div>
      </Fragment>
    );
  }

}


ChatList.propTypes = {
  chatLog: PropTypes.arrayOf(
    PropTypes.shape({
      messageId: PropTypes.string,
      userId: PropTypes.string,
      fullName: PropTypes.string,
      timestamp: PropTypes.string,
      email: PropTypes.string,
      message: PropTypes.string,
      avatar: PropTypes.string
    })
  )
};

ChatMessage.defaultProps = {
  chatLog: []
};


export default ChatList;