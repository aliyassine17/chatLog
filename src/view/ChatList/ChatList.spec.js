import React from 'react';
import { shallow } from 'enzyme';
import ChatList from './index';

describe('<ChatList>', ()=>{
  let wrapper = null;
  const chatLog = [{
    messageId: '12356',
    userId: '613651251',
    fullName: 'Robin Balmforth',
    timestamp: '2016-06-03T20:24:29Z',
    email: 'robin@example.com',
    message: 'Hello, World!',
    avatar: null
  },
  {
    messageId: '123567',
    userId: '6136512517',
    fullName: 'Robin Balmforthh',
    timestamp: '2016-06-03T20:24:29Zh',
    email: 'robin6@example.com',
    message: 'Hello, World!',
    avatar: null
  }];

  beforeEach(()=>{
    wrapper = shallow(<ChatList chatLog={chatLog}/>);
    jest.spyOn(ChatList.prototype, 'handleChange');
  });

  afterEach(()=>{
    wrapper = null;
  });

  it('renders without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have one messages container', () => {
    expect(wrapper.find('.messages').length).toEqual(1);
  });
  it('should have one filter', () => {
    expect(wrapper.find('.filter').length).toEqual(1);
  });
  it('should have 10 records to show by default ', () => {
    expect(wrapper.state().showRecords).toBe(10);
  });
  it('should simulate change and update the state ', () => {
    const target = {
      target :{
        name: 'showRecords',
        value: 20,
        id: 'showRecords'
      }
    };
    expect(wrapper.state().showRecords).toBe(10);
    wrapper.find('#showRecords').simulate('change', target);
    expect(ChatList.prototype.handleChange).toHaveBeenCalled();
    expect(wrapper.state().showRecords).toBe(20);
  });
  it('should load the chatlog into the View ', () => {

    expect(wrapper.state().showRecords).toBe(10);
    wrapper.setState({
      chatLog
    });
    expect(wrapper.state().chatLog.length).toBe(2);
  });
  it('should load 1 ChatMessage component ', () => {
    wrapper.setState({
      chatLog
    });
    expect(wrapper.find('ChatMessage').length).toBe(2);
  });
});