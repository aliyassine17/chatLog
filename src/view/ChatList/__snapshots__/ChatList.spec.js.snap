// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChatList> renders without crashing 1`] = `
<Fragment>
  <div
    className="filter"
  >
    <label
      htmlFor="showRecords"
    >
      Records per page 
    </label>
    <input
      id="showRecords"
      max={0}
      min="1"
      name="showRecords"
      onChange={[Function]}
      type="number"
      value={10}
    />
  </div>
  <div
    className="messages"
  />
</Fragment>
`;
